File: subprojects/docs/src/samples/userguide/plugins/buildService/java/buildSrc/src/main/java/Download.java
import org.gradle.api.DefaultTask;
import org.gradle.api.file.RegularFileProperty;
import org.gradle.api.provider.Property;
import org.gradle.api.tasks.Internal;
import org.gradle.api.tasks.OutputFile;
import org.gradle.api.tasks.TaskAction;

public abstract class Download extends DefaultTask {
    @Internal
    abstract Property<WebServer> getServer();

    @OutputFile
    abstract RegularFileProperty getOutputFile();

    @TaskAction
    public void download() {
        WebServer server = getServer().get();
        // use the server to download a file
    }
}

File: subprojects/docs/src/samples/userguide/plugins/buildService/kotlin/buildSrc/src/main/kotlin/Download.kt
import org.gradle.api.DefaultTask
import org.gradle.api.file.RegularFileProperty
import org.gradle.api.provider.Property
import org.gradle.api.tasks.Internal
import org.gradle.api.tasks.OutputFile
import org.gradle.api.tasks.TaskAction

abstract class Download : DefaultTask() {
    @get:Internal
    abstract val server: Property<WebServer>

    @get:OutputFile
    abstract val outputFile: RegularFileProperty

    @TaskAction
    fun download() {
        val server = server.get()
        // use the server to download a file
    }
}
